bug: flaky test in storage download ambiguous-prefix resolution

## Summary

The test `should return 400 when version prefix is ambiguous` in `app/api/storages/download/__tests__/route.test.ts` is flaky and intermittently fails in CI.

## Observed Behavior

**Test:** `GET /api/storages/download > version prefix resolution > should return 400 when version prefix is ambiguous`

**Expected:** Response contains `"Ambiguous"` error message  
**Actual (intermittent):** Response contains `"version: Version must be 8-64 hex characters"`

```
AssertionError: expected 'version: Version must be 8-64 hex cha…' to contain 'Ambiguous'

Expected: "Ambiguous"
Received: "version: Version must be 8-64 hex characters"

 ❯ app/api/storages/download/__tests__/route.test.ts:246:34
```

## Analysis

The test creates two storage versions that share the same 8-character prefix, then queries with that prefix expecting an "Ambiguous" error. However, intermittently:

1. The Zod validation rejects the request before reaching the business logic
2. This suggests the version prefix may occasionally contain non-hex characters or have incorrect length

**Potential causes:**
- Race condition in test setup (second version not committed before query)
- `versionId` generation producing unexpected format in certain conditions
- Test isolation issue with database state

## Evidence

- **Local run:** Test passes consistently
- **CI run 21815348876:** Test failed, then passed on rerun
- **Main branch:** Test passes (run 21815083784)

## Reproduction

The failure is intermittent and difficult to reproduce locally. It appears primarily in CI environment.

## Suggested Fix

1. Add explicit wait/verification that second version is committed before querying
2. Add debug logging to capture actual `versionId` and prefix values when test fails
3. Consider making the test deterministic by using fixed UUIDs

## Related

- Test file: `turbo/apps/web/app/api/storages/download/__tests__/route.test.ts:230-247`
- Previous fix attempt: #2572 (`fix(web): make storage download ambiguous-prefix test deterministic`)

--- a/turbo/apps/cli/src/commands/connector/connect.ts
+++ b/turbo/apps/cli/src/commands/connector/connect.ts
@@ -56,7 +56,7 @@ export const connectCommand = new Command()
       const sessionsClient = initClient(connectorSessionsContract, {
         baseUrl: apiUrl,
         baseHeaders: headers,
-        jsonQuery: true,
+        jsonQuery: false,
       });
 
       const createResult = await sessionsClient.create({
@@ -87,7 +87,7 @@ export const connectCommand = new Command()
       const sessionClient = initClient(connectorSessionByIdContract, {
         baseUrl: apiUrl,
         baseHeaders: headers,
-        jsonQuery: true,
+        jsonQuery: false,
       });
 
       const startTime = Date.now();

--- a/turbo/apps/cli/src/lib/api/api-client.ts
+++ b/turbo/apps/cli/src/lib/api/api-client.ts
@@ -162,7 +162,7 @@ class ApiClient {
     const client = initClient(composesMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getByName({
@@ -191,7 +191,7 @@ class ApiClient {
     const client = initClient(composesByIdContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getById({
@@ -221,11 +221,10 @@ class ApiClient {
     const headers = await this.getHeaders();
 
     // Create ts-rest client with config
-    // Note: jsonQuery: true handles scientific notation edge cases automatically
     const client = initClient(composesVersionsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.resolveVersion({
@@ -256,7 +255,7 @@ class ApiClient {
     const client = initClient(composesMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.create({
@@ -305,7 +304,7 @@ class ApiClient {
     const client = initClient(runsMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.create({ body });
@@ -333,7 +332,7 @@ class ApiClient {
     const client = initClient(runEventsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getEvents({
@@ -366,7 +365,7 @@ class ApiClient {
     const client = initClient(runSystemLogContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getSystemLog({
@@ -400,7 +399,7 @@ class ApiClient {
     const client = initClient(runMetricsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getMetrics({
@@ -434,7 +433,7 @@ class ApiClient {
     const client = initClient(runAgentEventsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getAgentEvents({
@@ -468,7 +467,7 @@ class ApiClient {
     const client = initClient(runNetworkLogsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getNetworkLogs({
@@ -502,7 +501,7 @@ class ApiClient {
     const client = initClient(scopeContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.get({ headers: {} });
@@ -532,7 +531,7 @@ class ApiClient {
     const client = initClient(scopeContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.create({ body });
@@ -562,7 +561,7 @@ class ApiClient {
     const client = initClient(scopeContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.update({ body });
@@ -590,7 +589,7 @@ class ApiClient {
     const client = initClient(sessionsByIdContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getById({
@@ -621,7 +620,7 @@ class ApiClient {
     const client = initClient(checkpointsByIdContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getById({
@@ -662,7 +661,7 @@ class ApiClient {
     const client = initClient(storagesPrepareContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.prepare({ body });
@@ -698,7 +697,7 @@ class ApiClient {
     const client = initClient(storagesCommitContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.commit({ body });
@@ -739,7 +738,7 @@ class ApiClient {
     const client = initClient(storagesDownloadContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.download({
@@ -777,7 +776,7 @@ class ApiClient {
     const client = initClient(storagesListContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ query });
@@ -813,7 +812,7 @@ class ApiClient {
     const client = initClient(schedulesMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.deploy({ body });
@@ -837,7 +836,7 @@ class ApiClient {
     const client = initClient(schedulesMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ headers: {} });
@@ -864,7 +863,7 @@ class ApiClient {
     const client = initClient(schedulesByNameContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.getByName({
@@ -895,7 +894,7 @@ class ApiClient {
     const client = initClient(schedulesByNameContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.delete({
@@ -927,7 +926,7 @@ class ApiClient {
     const client = initClient(schedulesEnableContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.enable({
@@ -958,7 +957,7 @@ class ApiClient {
     const client = initClient(schedulesEnableContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.disable({
@@ -990,7 +989,7 @@ class ApiClient {
     const client = initClient(scheduleRunsContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.listRuns({
@@ -1029,7 +1028,7 @@ class ApiClient {
     const client = initClient(publicAgentsListContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ query: query ?? {} });
@@ -1059,7 +1058,7 @@ class ApiClient {
     const client = initClient(publicArtifactsListContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ query: query ?? {} });
@@ -1083,7 +1082,7 @@ class ApiClient {
     const client = initClient(publicArtifactByIdContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.get({ params: { id } });
@@ -1113,7 +1112,7 @@ class ApiClient {
     const client = initClient(publicVolumesListContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ query: query ?? {} });
@@ -1137,7 +1136,7 @@ class ApiClient {
     const client = initClient(publicVolumeByIdContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.get({ params: { id } });
@@ -1189,7 +1188,7 @@ class ApiClient {
     const client = initClient(credentialsMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.list({ headers: {} });
@@ -1213,7 +1212,7 @@ class ApiClient {
     const client = initClient(credentialsByNameContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.get({
@@ -1244,7 +1243,7 @@ class ApiClient {
     const client = initClient(credentialsMainContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.set({ body });
@@ -1268,7 +1267,7 @@ class ApiClient {
     const client = initClient(credentialsByNameContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.delete({
@@ -1295,7 +1294,7 @@ class ApiClient {
     const client = initClient(realtimeTokenContract, {
       baseUrl,
       baseHeaders: headers,
-      jsonQuery: true,
+      jsonQuery: false,
     });
 
     const result = await client.create({

--- a/turbo/apps/cli/src/lib/api/core/client-factory.ts
+++ b/turbo/apps/cli/src/lib/api/core/client-factory.ts
@@ -56,11 +56,11 @@ export async function getBaseUrl(): Promise<string> {
 export async function getClientConfig(): Promise<{
   baseUrl: string;
   baseHeaders: Record<string, string>;
-  jsonQuery: true;
+  jsonQuery: false;
 }> {
   const baseUrl = await getBaseUrl();
   const baseHeaders = await getHeaders();
-  return { baseUrl, baseHeaders, jsonQuery: true };
+  return { baseUrl, baseHeaders, jsonQuery: false };
 }
 
 /**

--- a/turbo/apps/web/app/api/storages/download/__tests__/route.test.ts
+++ b/turbo/apps/web/app/api/storages/download/__tests__/route.test.ts
@@ -246,6 +246,34 @@ describe("GET /api/storages/download", () => {
       expect(json.error.message).toContain("Ambiguous");
     });
 
+    /**
+     * Guard test: version prefix that resembles scientific notation.
+     *
+     * jsonQuery is intentionally disabled in ts-rest-handler.ts (#2666).
+     * The hex prefix "846e3519" looks like scientific notation to JSON.parse(),
+     * which returns Infinity and corrupts the value before Zod validation.
+     * If someone re-enables jsonQuery, this test will fail.
+     */
+    it("should resolve version prefix that resembles scientific notation (jsonQuery guard)", async () => {
+      const storageName = uniqueId("sci-notation");
+      const sciNotationPrefix = "846e3519";
+      const fullVersionId = sciNotationPrefix + "b".repeat(56);
+
+      // Create storage without committing, then insert a version with the known prefix
+      await createTestArtifact(storageName, { skipCommit: true });
+      await insertStorageVersion(storageName, fullVersionId);
+
+      const request = createTestRequest(
+        `http://localhost:3000/api/storages/download?name=${storageName}&type=artifact&version=${sciNotationPrefix}`,
+      );
+
+      const response = await GET(request);
+      expect(response.status).toBe(200);
+
+      const json = await response.json();
+      expect(json.versionId).toBe(fullVersionId);
+    });
+
     it("should return 400 when version prefix contains invalid characters", async () => {
       const storageName = uniqueId("prefix-invalid");
       const files = [{ path: "test.txt", hash: "0".repeat(64), size: 100 }];

--- a/turbo/apps/web/src/lib/public-api/handler.ts
+++ b/turbo/apps/web/src/lib/public-api/handler.ts
@@ -55,7 +55,9 @@ export function createPublicApiHandler<T extends AppRouter>(
 ) {
   return createNextHandler(contract, router, {
     handlerType: "app-router",
-    jsonQuery: true,
+    // jsonQuery is intentionally disabled: JSON.parse() misinterprets hex strings
+    // like "846e3519" as scientific notation, corrupting version query params (#2666)
+    jsonQuery: false,
     errorHandler: options?.errorHandler ?? publicApiErrorHandler,
     requestMiddleware: [
       (request) => {

--- a/turbo/apps/web/src/lib/ts-rest-handler.ts
+++ b/turbo/apps/web/src/lib/ts-rest-handler.ts
@@ -98,7 +98,9 @@ export function createHandler<T extends AppRouter>(
 ) {
   return createNextHandler(contract, router, {
     handlerType: "app-router",
-    jsonQuery: true,
+    // jsonQuery is intentionally disabled: JSON.parse() misinterprets hex strings
+    // like "846e3519" as scientific notation, corrupting version query params (#2666)
+    jsonQuery: false,
     ...options,
     requestMiddleware: [
       (request) => {

--- a/turbo/packages/core/src/contracts/__tests__/version-query-schema.test.ts
+++ b/turbo/packages/core/src/contracts/__tests__/version-query-schema.test.ts
@@ -3,32 +3,23 @@ import { z } from "zod";
 
 /**
  * Version ID query parameter schema (copied from storages.ts for testing)
- *
- * Handles jsonQuery edge case where hex strings like "846e3519"
- * are parsed as JavaScript scientific notation numbers.
  */
-const versionQuerySchema = z.preprocess(
-  (val) => (val === undefined || val === null ? undefined : String(val)),
-  z
-    .string()
-    .regex(/^[a-f0-9]{8,64}$/i, "Version must be 8-64 hex characters")
-    .optional(),
-);
+const versionQuerySchema = z
+  .string()
+  .regex(/^[a-f0-9]{8,64}$/i, "Version must be 8-64 hex characters")
+  .optional();
 
 /**
  * Compose version query schema (copied from composes.ts for testing)
  * Also accepts "latest" tag
  */
-const composeVersionQuerySchema = z.preprocess(
-  (val) => (val === undefined || val === null ? undefined : String(val)),
-  z
-    .string()
-    .min(1, "Missing version query parameter")
-    .regex(
-      /^[a-f0-9]{8,64}$|^latest$/i,
-      "Version must be 8-64 hex characters or 'latest'",
-    ),
-);
+const composeVersionQuerySchema = z
+  .string()
+  .min(1, "Missing version query parameter")
+  .regex(
+    /^[a-f0-9]{8,64}$|^latest$/i,
+    "Version must be 8-64 hex characters or 'latest'",
+  );
 
 describe("versionQuerySchema (storages)", () => {
   describe("valid inputs", () => {
@@ -46,10 +37,6 @@ describe("versionQuerySchema (storages)", () => {
       expect(versionQuerySchema.parse(undefined)).toBeUndefined();
     });
 
-    it("should accept null (converts to undefined)", () => {
-      expect(versionQuerySchema.parse(null)).toBeUndefined();
-    });
-
     it("should accept uppercase hex characters", () => {
       expect(versionQuerySchema.parse("ABCD1234")).toBe("ABCD1234");
     });
@@ -58,56 +45,17 @@ describe("versionQuerySchema (storages)", () => {
       expect(versionQuerySchema.parse("AbCd1234")).toBe("AbCd1234");
     });
 
-    it("should handle scientific notation hex strings correctly", () => {
-      // This is the exact case that caused the flaky test
-      // "846e3519" looks like scientific notation to JSON.parse
+    it("should accept hex strings that resemble scientific notation", () => {
+      // "846e3519" looks like scientific notation to JSON.parse but is valid hex
       expect(versionQuerySchema.parse("846e3519")).toBe("846e3519");
     });
 
-    it("should handle another scientific notation pattern", () => {
+    it("should accept another scientific notation pattern", () => {
       expect(versionQuerySchema.parse("123e4567")).toBe("123e4567");
     });
   });
 
-  describe("invalid inputs - numbers from JSON.parse", () => {
-    it("should convert Infinity (from JSON.parse) to string and reject", () => {
-      // JSON.parse("846e3519") returns Infinity
-      expect(() => versionQuerySchema.parse(Infinity)).toThrow(
-        "Version must be 8-64 hex characters",
-      );
-    });
-
-    it("should reject scientific notation number (the flaky test root cause)", () => {
-      // JSON.parse("52999e37") returns 5.2999e+41 (not Infinity, but a large number)
-      // String(5.2999e+41) = "5.2999e+41" which is NOT valid hex
-      // This is the exact bug that caused the flaky cli-e2e-02-parallel test
-      const parsedNumber = JSON.parse("52999e37"); // 5.2999e+41
-      expect(() => versionQuerySchema.parse(parsedNumber)).toThrow(
-        "Version must be 8-64 hex characters",
-      );
-    });
-
-    it("should convert plain number to string - valid if 8+ hex digits", () => {
-      // JSON.parse("12345678") returns 12345678
-      // String(12345678) = "12345678" which IS valid hex (0-9 are hex chars)
-      expect(versionQuerySchema.parse(12345678)).toBe("12345678");
-    });
-
-    it("should convert short number to string and reject", () => {
-      // String(1234567) = "1234567" (7 chars, too short)
-      expect(() => versionQuerySchema.parse(1234567)).toThrow(
-        "Version must be 8-64 hex characters",
-      );
-    });
-
-    it("should convert NaN to string and reject", () => {
-      expect(() => versionQuerySchema.parse(NaN)).toThrow(
-        "Version must be 8-64 hex characters",
-      );
-    });
-  });
-
-  describe("invalid inputs - bad format", () => {
+  describe("invalid inputs", () => {
     it("should reject strings shorter than 8 chars", () => {
       expect(() => versionQuerySchema.parse("abc1234")).toThrow(
         "Version must be 8-64 hex characters",
@@ -155,41 +103,12 @@ describe("composeVersionQuerySchema (composes)", () => {
       expect(composeVersionQuerySchema.parse("LATEST")).toBe("LATEST");
     });
 
-    it("should handle scientific notation hex strings correctly", () => {
+    it("should accept hex strings that resemble scientific notation", () => {
       expect(composeVersionQuerySchema.parse("846e3519")).toBe("846e3519");
     });
   });
 
-  describe("invalid inputs - numbers from JSON.parse", () => {
-    it("should convert Infinity to string and reject", () => {
-      expect(() => composeVersionQuerySchema.parse(Infinity)).toThrow(
-        "Version must be 8-64 hex characters or 'latest'",
-      );
-    });
-
-    it("should reject scientific notation number (the flaky test root cause)", () => {
-      // JSON.parse("52999e37") returns 5.2999e+41 (not Infinity, but a large number)
-      // String(5.2999e+41) = "5.2999e+41" which is NOT valid hex
-      // This is the exact bug that caused the flaky cli-e2e-02-parallel test
-      const parsedNumber = JSON.parse("52999e37"); // 5.2999e+41
-      expect(() => composeVersionQuerySchema.parse(parsedNumber)).toThrow(
-        "Version must be 8-64 hex characters or 'latest'",
-      );
-    });
-
-    it("should convert plain number to string - valid if 8+ hex digits", () => {
-      // String(12345678) = "12345678" which IS valid hex
-      expect(composeVersionQuerySchema.parse(12345678)).toBe("12345678");
-    });
-
-    it("should convert short number to string and reject", () => {
-      expect(() => composeVersionQuerySchema.parse(1234567)).toThrow(
-        "Version must be 8-64 hex characters or 'latest'",
-      );
-    });
-  });
-
-  describe("invalid inputs - bad format", () => {
+  describe("invalid inputs", () => {
     it("should reject undefined (required field)", () => {
       expect(() => composeVersionQuerySchema.parse(undefined)).toThrow();
     });

--- a/turbo/packages/core/src/contracts/composes.ts
+++ b/turbo/packages/core/src/contracts/composes.ts
@@ -8,21 +8,15 @@ const c = initContract();
 /**
  * Version query parameter schema for compose versions
  *
- * Handles jsonQuery edge case where hex strings like "846e3519"
- * are parsed as JavaScript scientific notation numbers (e.g., 846e3519 = Infinity).
- *
  * Accepts: "latest" tag or 8-64 hex character version hash
  */
-const composeVersionQuerySchema = z.preprocess(
-  (val) => (val === undefined || val === null ? undefined : String(val)),
-  z
-    .string()
-    .min(1, "Missing version query parameter")
-    .regex(
-      /^[a-f0-9]{8,64}$|^latest$/i,
-      "Version must be 8-64 hex characters or 'latest'",
-    ),
-);
+const composeVersionQuerySchema = z
+  .string()
+  .min(1, "Missing version query parameter")
+  .regex(
+    /^[a-f0-9]{8,64}$|^latest$/i,
+    "Version must be 8-64 hex characters or 'latest'",
+  );
 
 /**
  * Agent name validation schema

--- a/turbo/packages/core/src/contracts/storages.ts
+++ b/turbo/packages/core/src/contracts/storages.ts
@@ -12,19 +12,12 @@ const storageTypeSchema = z.enum(["volume", "artifact"]);
 /**
  * Version ID query parameter schema
  *
- * Handles jsonQuery edge case where hex strings like "846e3519"
- * are parsed as JavaScript scientific notation numbers (e.g., 846e3519 = Infinity).
- *
- * The preprocess step converts any non-null/undefined value to string,
- * then validates it matches hex format (8-64 characters).
+ * Validates hex format (8-64 characters). Used for version prefix resolution.
  */
-const versionQuerySchema = z.preprocess(
-  (val) => (val === undefined || val === null ? undefined : String(val)),
-  z
-    .string()
-    .regex(/^[a-f0-9]{8,64}$/i, "Version must be 8-64 hex characters")
-    .optional(),
-);
+const versionQuerySchema = z
+  .string()
+  .regex(/^[a-f0-9]{8,64}$/i, "Version must be 8-64 hex characters")
+  .optional();
 
 /**
  * Upload storage response schema (JSON part of POST response)

